app/src/main/java/com/nutomic/syncthingandroid/util/ConfigXml.java

```diff
@@ -110,9 +124,65 @@ public String getUserName() {
     @SuppressWarnings("SdCardPath")
     public void updateIfNeeded() {
-        Log.i(TAG, "Checking for needed config updates");
         boolean changed = false;
+
+        /* Get preference - PREF_USE_FOLDER_OBSERVER */
+        mPreferences = PreferenceManager.getDefaultSharedPreferences(mContext);
+        boolean prefUseFolderObserver = mPreferences.getBoolean(Constants.PREF_USE_FOLDER_OBSERVER, false);
+
+        /* Read existing config version */
+        int iConfigVersion = Integer.parseInt(mConfig.getDocumentElement().getAttribute("version"));
+        Log.i(TAG, "Found existing config version " + Integer.toString(iConfigVersion));
+
+        /* Get refs to important config objects */
         NodeList folders = mConfig.getDocumentElement().getElementsByTagName("folder");
+
+        /* Check if we have to do manual migration from version X to Y */
+        if (iConfigVersion == 27) {
+          /* fsWatcher transition - https://github.com/syncthing/syncthing/issues/4882 */
+          Log.i(TAG, "Migrating config version " + Integer.toString(iConfigVersion) + " to 28 ...");
+
+          /* Enable fsWatcher for all folders */
+          for (int i = 0; i < folders.getLength(); i++) {
+            Element r = (Element) folders.item(i);
+
+            // Enable "fsWatcherEnabled" attribute.
+            Log.i(TAG, "Set 'fsWatcherEnabled' on folder " + r.getAttribute("id"));
+            r.setAttribute("fsWatcherEnabled", Boolean.toString(true));
+            changed = true;
+          }
+
+          /**
+            * Set config version to 28 after manual config migration
+            * This prevents "unackedNotificationID" getting populated
+            * with the fsWatcher GUI notification.
+            */
+          iConfigVersion = 28;
+          mConfig.getDocumentElement().setAttribute("version", Integer.toString(iConfigVersion));
+          Log.i(TAG, "Config version " + Integer.toString(iConfigVersion) + " reached.");
+        }
+
+        /**
+          * Force-disable fsWatcher for all folders if prefUseFolderObserver has been manually set
+          * in experimental options. This should give users the option to go back to the legacy
+          * implementation with FolderObserver watching for file changes instead of fsWatcher.
+          * Intended to be advised in the issue tracker if a user encounters a critical bug with
+          * the new fsWatcher. To be removed in a future release.
+        */
+        if (prefUseFolderObserver) {
+          Log.i(TAG, "Disabling fsWatcher on all folders because experimental option to use FolderObserver was manually set.");
+          for (int i = 0; i < folders.getLength(); i++) {
+            Element r = (Element) folders.item(i);
+
+            // Disable "fsWatcherEnabled" attribute.
+            if (r.hasAttribute("fsWatcherEnabled")) {
+              r.setAttribute("fsWatcherEnabled", Boolean.toString(false));
+              changed = true;
+            }
+          }
+        }
```

Comment:
Your formatting is off compared to existing code (blocks should be indented by 4 spaces not 2). Same thing in other places in this PR. There should be a setting for this in your editor.

Reply:
Thanks for taking the time. I corrected intendations and moved the migration code to a separate function.